docs: add guide for CI/CD template testing (#15528)
coder/coder: docs/tutorials/testing-templates.md

    runs-on: ubuntu-latest
    env:
      TEMPLATE_NAME: "my-template"
    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Set up Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest

      - name: Set up Coder CLI
        uses: coder/setup-action@v1
        with:
          access_url: "https://coder.example.com"
          coder_session_token: ${{ secrets.CODER_SESSION_TOKEN }}

      - name: Validate Terraform template
        run: terraform validate

      - name: Get short commit SHA to use as template version name
        id: name
        run: echo "version_name=$(git rev-parse --short HEAD)" >> $GITHUB_OUTPUT

      - name: Get latest commit title to use as template version description
        id: message
        run:
          echo "pr_title=$(git log --format=%s -n 1 ${{ github.sha }})" >>
          $GITHUB_OUTPUT

      - name: Push template to Coder
        run: |
          coder templates push $TEMPLATE_NAME --activate=false --name ${{ steps.name.outputs.version_name }} --message "${{ steps.message.outputs.pr_title }}" --yes

      - name: Create a test workspace
        run: |

>           coder create -t $TEMPLATE_NAME --template-version ${{ steps.name.outputs.version_name }} test-${{ steps.name.outputs.version_name }} --yes

Reviewer: Should we delete the test workspace?

Author: Yes. Good catch. I will do.

    runs-on: ubuntu-latest
    env:
      TEMPLATE_NAME: "my-template"
    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Set up Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest

      - name: Set up Coder CLI
        uses: coder/setup-action@v1
        with:
          access_url: "https://coder.example.com"
          coder_session_token: ${{ secrets.CODER_SESSION_TOKEN }}

      - name: Validate Terraform template
        run: terraform validate

      - name: Get short commit SHA to use as template version name
        id: name
        run: echo "version_name=$(git rev-parse --short HEAD)" >> $GITHUB_OUTPUT

      - name: Get latest commit title to use as template version description
        id: message
        run:
          echo "pr_title=$(git log --format=%s -n 1 ${{ github.sha }})" >>
          $GITHUB_OUTPUT

      - name: Push template to Coder
        run: |
          coder templates push $TEMPLATE_NAME --activate=false --name ${{ steps.name.outputs.version_name }} --message "${{ steps.message.outputs.pr_title }}" --yes

      - name: Create a test workspace
        run: |

>           coder create -t $TEMPLATE_NAME --template-version ${{ steps.name.outputs.version_name }} --yes

Reviewer: Can we also show how to test if specific conditions work inside the template (e.g. `make build`) or something? Like how to run example commands and ensure they run properly? The `coder exp scaletest` command may be useful for this.

> # Testing and Publishing Coder Templates in CI/CD

Reviewer 2: ```suggestion
# Test and Publish Coder Templates Through CI/CD
```

<div>
  <a href="https://github.com/matifali" style="text-decoration: none; color: inherit;">
    <span style="vertical-align:middle;">Muhammad Atif Ali</span>
    <img src="https://github.com/matifali.png" width="24px" height="24px" style="vertical-align:middle; margin: 0px;"/>
  </a>
</div>
November 15, 2024

---

## Overview

This guide demonstrates how to test and publish Coder templates in a Continuous
Integration (CI) pipeline using the
[coder/setup-action](https://github.com/coder/setup-coder). This workflow
ensures your templates are validated, tested, and promoted seamlessly.

## Prerequisites

> Before proceeding, ensure the following:
> 
> - **Coder CLI** is installed and configured in your environment.
> - **Terraform CLI** is installed and available in your CI environment.
> - Access to your **Coder instance** with the appropriate
>   [permissions](../admin/users/groups-roles.md#roles).
> 

Reviewer 2: ```suggestion
- Install and configure Coder CLI in your environment.
- Install Terraform CLI in your CI environment.
- Configure [user roles and permissions](../admin/users/groups-roles.md#roles) for your Coder instance.

```

## Example GitHub Action Workflow

> Below is an example workflow for testing and publishing a template using GitHub
> Actions. The workflow first validates the Terraform template, pushes the
> template to Coder without activating it, tests the template by creating a
> workspace, and then promotes the template version to active upon successful
> workspace creation.
> 
> ### Step-by-Step Process
> 
> 1. **Validate the Terraform template.**
> 2. **Push the template to Coder without activating it.**
> 3. **Test the template by creating a workspace.**
> 4. **Promote the template version to active upon successful workspace
>    creation.**

Reviewer 2: ```suggestion
This example workflow tests and publishes a template using GitHub
Actions.

The workflow:

1. Validates the Terraform template.
1. Pushes the template to Coder without activating it.
1. Tests the template by creating a workspace.
1. Promotes the template version to active upon successful workspace creation.
```
